MAINT: Clean-up uses of np.asarray in ndimage

jaimefrio opened the pull request:
Originally about avoiding unnecessary copies being made by calling `.astype()` afterwards, rather than including the desired `dtype` in the call to `asarray()`.

Have also included removing some unreachable code paths and some unnecessary checks.

jaimefrio commented on scipy/ndimage/morphology.py:
```diff
@@ -257,11 +253,7 @@ def _binary_erosion(input, structure, iterations, mask, output,
             if not structure.shape[ii] & 1:
                 origin[ii] -= 1
         if mask is not None:
-            msk = numpy.asarray(mask)
```
This whole branch made very little sense: calling `.astype()` always makes a copy, so the check for `msk is mask` was completely unnecessary. It is now possible that mask is the original array, not a copy, but `binary_erosion2()` never writes to mask, only reads it, and it will make a copy if the array is not aligned and contiguous.

jaimefrio commented on scipy/ndimage/morphology.py:
```diff
@@ -205,13 +205,10 @@ def generate_binary_structure(rank, connectivity):
     output = numpy.fabs(numpy.indices([3] * rank) - 1)
     output = numpy.add.reduce(output, 0)
-    return numpy.asarray(output <= connectivity, dtype=bool)
+    return output <= connectivity
```
`output` is always an `ndarray`, so a comparison on it will always be a boolean array, the extra call was unnecessary.

perimosocordiae commented on scipy/ndimage/filters.py:
```diff
@@ -1066,8 +1066,7 @@ def _min_or_max_filter(input, size, footprint, structure, output, mode,
         else:
-            footprint = numpy.asarray(footprint)
-            footprint = footprint.astype(bool)
+            footprint = numpy.asarray(footprint, dtype=bool)
```
I'm not as familiar with ndimage, but I assume that `footprint` is never modified by the native `min_or_max_filter` code? Otherwise, this change could lead to an in-place modification bug.

jaimefrio replied:
No, it is only read, never written to.

perimosocordiae commented on scipy/ndimage/morphology.py:
```diff
@@ -205,13 +205,10 @@ def generate_binary_structure(rank, connectivity):
-            return numpy.array(0, dtype=bool)
-        else:
-            return numpy.array(1, dtype=bool)
+        return numpy.array(1, dtype=bool)
```
`numpy.array(True, dtype=bool)`

jaimefrio replied:
Done!

jaimefrio commented on scipy/ndimage/filters.py:
```diff
@@ -1066,8 +1066,7 @@ def _min_or_max_filter(input, size, footprint, structure, output, mode,
+            footprint = numpy.asarray(footprint, dtype=bool)
             if not footprint.any():
                 raise ValueError("All-zero footprint is not supported.")
             if numpy.alltrue(numpy.ravel(footprint), axis=0):
```
It is, have added that change to the PR as well.

jaimefrio commented:
Have addressed the review comments. I actually come accross this by searching for all uses of `asarray` in `ndimage`. There are more uses, but many are properly including the dtype in the call to `asarray`.

perimosocordiae commented:
I'm +1 to merge once the CI tests come in.

rgommers commented:
Tests are happy apart from the timeout, so in it goes. Thanks @jaimefrio, @perimosocordiae
